protocols/v2/roles-logic-sv2/src/utils.rs

```diff
@@ -276,48 +280,94 @@ impl TryFrom<CoinbaseOutput> for Script {
 /// [3] https://en.wikipedia.org/wiki/Negative_hypergeometric_distribution
 /// bdiff: 0x00000000ffff0000000000000000000000000000000000000000000000000000
 /// https://en.bitcoin.it/wiki/Difficulty#How_soon_might_I_expect_to_generate_a_block.3F
-pub fn hash_rate_to_target(h: f32, share_per_min: f32) -> U256<'static> {
+
+pub fn hash_rate_to_target(
+    hashrate: f32,
+    share_per_min: f32,
+) -> Result<U256<'static>, crate::Error> {
+    // checks that we are not dividing by zero
+    if share_per_min == 0.0 {
+        return Err(Error::ImpossibleToGetTarget);
+    }
+
     // if we want 5 shares per minute, this means that s=60/5=12 seconds interval between shares
-    let s: f32 = 60_f32 / share_per_min;
-    let h_times_s = (h * s) as u128;
+    // this quantity will be at the numerator, so we multiply the result by 100 again later
+    let shares_occurrency_frequence = 60_f32 / share_per_min;
+    let shares_occurrency_frequence = shares_occurrency_frequence;
 
+    if hashrate.is_sign_negative() {
+        return Err(Error::ImpossibleToGetTarget);
```

Comment:
Same her use a more decsriptive error. Also input validators should be at the beginning of the function.